### PR TITLE
Refactor to add all citizens in general to the abstractBuilding

### PR DESCRIPTION
**Pull request by Raycoms**
Avoid maintaining two lists!

Refactors the building code to put all citizen handling code into 1 class.
(Less duplicated code, easier to maintain)

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java [hunk 1]
```diff
@@ -1328,14 +1345,161 @@ private ItemStack forceItemStackToProvider(@NotNull final ICapabilityProvider pr
...
+    }
+
+    /**
+     * Assign the citizen to the current building.
+     *
+     * @param citizen {@link CitizenData} of the worker
+     */
+    public boolean assignCitizen(final CitizenData citizen)
+    {
+        if (assignedCitizen.contains(citizen))
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java:1442]: MINOR This call to "contains()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java [hunk 2]
```diff
@@ -1328,14 +1345,161 @@ private ItemStack forceItemStackToProvider(@NotNull final ICapabilityProvider pr
...
+
+    /**
+     * Returns whether the citizen has this as home or not.
+     *
+     * @param citizen Citizen to check.
+     * @return True if citizen lives here, otherwise false.
+     */
+    public boolean hasAssignedCitizen(final CitizenData citizen)
+    {
+        return assignedCitizen.contains(citizen);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java:1476]: MINOR This call to "contains()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java [hunk 3]
```diff
@@ -18,12 +18,14 @@
 import com.minecolonies.api.crafting.ItemStorage;
 import com.minecolonies.api.util.*;
 import com.minecolonies.api.util.constant.TypeConstants;
+import com.minecolonies.coremod.MineColonies;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java:21]: MINOR Remove this unused import 'com.minecolonies.coremod.MineColonies'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java [hunk 4]
```diff
@@ -1328,14 +1345,161 @@ private ItemStack forceItemStackToProvider(@NotNull final ICapabilityProvider pr
...
+
+    /**
+     * Returns if the {@link CitizenData} is the same as the worker.
+     *
+     * @param citizen {@link CitizenData} you want to compare
+     * @return true if same citizen, otherwise false
+     */
+    public boolean isCitizenAssigned(final CitizenData citizen)
+    {
+        return assignedCitizen.contains(citizen);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java:1397]: MINOR This call to "contains()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java [hunk 5]
```diff
@@ -1328,14 +1345,161 @@ private ItemStack forceItemStackToProvider(@NotNull final ICapabilityProvider pr
...
+    /**
+     * Method to remove a citizen.
+     *
+     * @param citizen Citizen to be removed.
+     */
+    public void removeCitizen(final CitizenData citizen)
+    {
+        if (isCitizenAssigned(citizen))
+        {
+            assignedCitizen.remove(citizen);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java:1383]: MINOR This call to "remove()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java [hunk 6]
```diff
@@ -25,33 +26,19 @@
...
 import static com.minecolonies.api.util.constant.ColonyConstants.NUM_ACHIEVEMENT_FIRST;
 import static com.minecolonies.api.util.constant.Constants.MAX_BUILDING_LEVEL;
+import static com.minecolonies.api.util.constant.NbtTagConstants.TAG_BEDS;
+import static com.minecolonies.api.util.constant.NbtTagConstants.TAG_RESIDENTS;
 
 /**
  * The class of the citizen hut.
  */
-public class BuildingHome extends AbstractBuildingHut
+public class BuildingHome extends AbstractBuilding
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java:35]: MINOR Override the "equals" method in this class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2160)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java [hunk 7]
```diff
@@ -314,21 +286,10 @@ public void setBuildingLevel(final int level)
...
-    public boolean hasResident(final CitizenData citizen)
-    {
-        return residents.contains(citizen);
-    }
-
     /**
      * The view of the citizen hut.
      */
-    public static class View extends AbstractBuildingHut.View
+    public static class View extends AbstractBuildingView
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingHome.java:292]: MAJOR Reduce this class from 38 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/colony/managers/StatisticAchievementManager.java [hunk 8]
```diff
@@ -244,7 +244,7 @@ private void incrementStatisticAmount(@NotNull final String statistic)
     public void checkAchievements()
     {
         // the colonies size
-        //final int size = this.citizenManager.getCitizens().size();
+        //final int size = this.citizenManager.getAssignedCitizen().size();
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/managers/StatisticAchievementManager.java:247]: MINOR This block of commented-out lines of code should be removed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: src/main/java/com/minecolonies/coremod/colony/managers/StatisticAchievementManager.java [hunk 9]
```diff
@@ -244,7 +244,7 @@ private void incrementStatisticAmount(@NotNull final String statistic)
...
 
         //todo check those later again
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/managers/StatisticAchievementManager.java:249]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/managers/StatisticAchievementManager.java [hunk 10]
```diff
@@ -244,7 +244,7 @@ private void incrementStatisticAmount(@NotNull final String statistic)
...
         /*if (size >= ModAchievements.ACHIEVEMENT_SIZE_SETTLEMENT)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/managers/StatisticAchievementManager.java:250]: MINOR This block of commented-out lines of code should be removed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 11]
```diff
@@ -38,44 +38,25 @@
...
 
+import static com.minecolonies.api.util.constant.NbtTagConstants.*;
 import static com.minecolonies.api.util.constant.ToolLevelConstants.TOOL_LEVEL_MAXIMUM;
 import static com.minecolonies.api.util.constant.ToolLevelConstants.TOOL_LEVEL_WOOD_OR_GOLD;
 
 /**
  * The abstract class for each worker building.
  */
-public abstract class AbstractBuildingWorker extends AbstractBuildingHut
+public abstract class AbstractBuildingWorker extends AbstractBuilding
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:48]: MINOR Override the "equals" method in this class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2160)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 12]
```diff
@@ -465,18 +368,28 @@ public void onWorldTick(@NotNull final TickEvent.WorldTickEvent event)
 
         // If we have no active worker, grab one from the Colony
         // TODO Maybe the Colony should assign jobs out, instead?
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:370]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 13]
```diff
@@ -575,7 +474,7 @@ else if (getBuildingLevel() <= WOOD_HUT_LEVEL)
     /**
      * AbstractBuildingWorker View for clients.
      */
-    public static class View extends AbstractBuildingHut.View
+    public static class View extends AbstractBuildingView
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:477]: MAJOR Reduce this class from 83 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java [hunk 14]
```diff
@@ -117,7 +117,7 @@ public String getJobName()
     @Override
     public void searchWorkOrder()
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/BuildingBuilder.java:118]: MAJOR Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingStructureBuilder.java [hunk 15]
```diff
@@ -194,8 +194,8 @@ public void serializeToView(@NotNull final ByteBuf buf)
      */
     private void updateAvailableResources()
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingStructureBuilder.java:195]: MAJOR Refactor this method to reduce its Cognitive Complexity from 22 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java [hunk 16]
```diff
@@ -205,7 +205,7 @@ public void onUpgradeComplete(final int newLevel)
     {
         if (newLevel > MAX_VISION_BONUS_MULTIPLIER)
         {
-            getWorkerEntities().stream()
+            getAssignedEntities().stream()
             .filter(Optional::isPresent)
             .map(Optional::get)
               .forEach(entityCitizen -> entityCitizen.getEntityAttribute(SharedMonsterAttributes.MAX_HEALTH).setBaseValue(SharedMonsterAttributes.MAX_HEALTH.getDefaultValue() + getBonusHealth()));
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingGuards.java:211]: MAJOR Split this 196 characters long line (which is greater than 180 authorized). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00103)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 112 issues
* ![CRITICAL][CRITICAL] 7 critical
* ![MAJOR][MAJOR] 72 major
* ![MINOR][MINOR] 31 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L237](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L237): Remove this call from a constructor to the overridable "setupRsDataStore" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1530](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1530): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1600](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1600): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1631](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1631): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1643](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1643): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1746](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1746): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuildingWorker.java#L69](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L69): Remove this call from a constructor to the overridable "getBuildingLevel" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/7781c76c6def1de2f07bdaf24fb921840a176fca/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — DISMISSED** [review]

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 112 issues
* ![CRITICAL][CRITICAL] 7 critical
* ![MAJOR][MAJOR] 72 major
* ![MINOR][MINOR] 31 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L237](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L237): Remove this call from a constructor to the overridable "setupRsDataStore" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1530](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1530): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1600](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1600): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1631](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1631): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1643](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1643): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuilding.java#L1746](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java#L1746): Remove usage of generic wildcard type. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)
1. ![CRITICAL][CRITICAL] [AbstractBuildingWorker.java#L69](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java#L69): Remove this call from a constructor to the overridable "getBuildingLevel" method. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1699)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/525acf82e85d381cf3fac99757d7a3b0d45032e9/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**